qmulticast/utils/multipartite_network.py

```diff
@@ -0,0 +1,207 @@
+    logger.debug("Unpacking edges.")
+    
+    edges = {}
+    for edge in graph.edges.data():
+        start, stop, weight = edge[0], edge[1], edge[2].get('weight', 1)
+        if str(start) != node.name:
+            continue
+        if type(weight) not in [int, float]:
+            raise TypeError("Edge weights must be numeric.")
+        edges[stop] = weight
+
+    return edges
+        
+def create_multipartite_network(name: str, graph: DiGraph) -> Network:
+    """Turn graph into netsquid network.
+
+    Give each node a bipatite source for each edge, assign memory
+    size and redirect to memory slots from connection ports.
+
+    Parameters
+    ----------
+    name : str
+        The name of the network.
+    graph : Graph
+        Graph representing the desired network.
+
+    Returns
+    -------
+    Network
+        A netsquid Network object.
+    """
+    logger.debug("Creating Network.")
+
+    # First set up NetSquid node objects for each graph node.
+    nodes = {node_name: Node(str(node_name)) for node_name in graph.nodes}
+
+    # Delay models to use for components.
+    source_delay = FixedDelayModel(delay=0)
+    fibre_delay = FibreDelayModel()
+
+    # Set up a state sampler for the |B00> bell state.
+    #state_sampler = StateSampler([ks.b00], [1.0])
+    state_sampler = StateSampler(gen_GHZ_ket(len(nodes)))
+    # Noise models to use for components.
+    # TODO find a suitable rate
+    depolar_noise = None  # DepolarNoiseModel(depolar_rate=1e-21)
+    source_noise = depolar_noise
+    # TODO do we want to change the default values?
+    fibre_loss = None  # FibreLossModel()
+
+    # Set up a Network object
+    network = Network(name=name)
+    logger.debug("Adding nodes to network.")
+    network.add_nodes([n for n in nodes.values()])
+
+    # Add unique components to each node
+    logger.debug("Adding unique components to nodes.")
+    for node_name, node in nodes.items():
+
+        # node_name = node.name
+        logger.debug(f"Node: {node_name}.")
+
+        node_connections = unpack_edge_values(node, graph)
+
+
+        # Names need to be strings for NetSquid object names
+        node_name = str(node_name)
+        # TODO intialise with routing chart
+        # TODO initalise given memory with respect to routing chart
+        # for now just assume one source and rest are max 1 edge away (??)
+        # give all maximium mem required e.g. hold whole GHZ state
+        mem_size = len(graph.nodes)*2
+
+        # Add a quantum memory to each of the nodes.
+        # TODO how much memory do we want to give?
+        # TODO change this to a processor as in tutorial "A full simulation"
+        logger.debug(f"Adding quantum memory 'qmemory-{node_name}'")
+        logger.debug(f"\tsize: {mem_size}")
+        qmemory = QuantumProcessor(
+            name="qmemory",
+            num_positions=mem_size,
+            memory_noise_models=depolar_noise,
+        )
+
+        node.add_subcomponent(qmemory)
```

Comment:
everything up to here is the same so reuse the bipartite bit by using one function. Probably easiest to have this main network function call out to a `bipartite` or `multipartite` part to add sources as needed. Assuming everything else is similar.

Reply:
I've started this my moving some of the constructor into its own function, could combine them and have they just call their specific subroutines

Reply:
I get what you're saying about the main function doing it but haven't done that yet